app/src/main/java/org/openobservatory/ooniprobe/activity/BrowserActivity.java

```diff
@@ -0,0 +1,275 @@
+            super.onPageFinished(view, url);
+            refresh.setVisibility(View.VISIBLE);
+            if (webViewProgressBar.isShown())
+                webViewProgressBar.setVisibility(View.GONE);
+            System.out.println("Finished loading " + url);
+            if (url.substring(0, 5).equals("https"))
+                urlLabel.setText("\uD83D\uDD12 " + url);
+            else
+                urlLabel.setText(url);
+            reloadButtons();
+        }
+
+        //TODO keep the error toasts for all messages? are they necessary?
+        @Override
+        public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
+            super.onReceivedError(view, request, error);
+            refresh.setVisibility(View.VISIBLE);
+            if (webViewProgressBar.isShown())
+                webViewProgressBar.setVisibility(View.GONE);
+            Toast.makeText(BrowserActivity.this, "Unexpected error occurred.Reload page again.", Toast.LENGTH_SHORT).show();
+        }
+
+        //TODO this should be removed includes error in internal frames
+        @Override
+        public void onReceivedHttpError(WebView view, WebResourceRequest request, WebResourceResponse errorResponse) {
+            super.onReceivedHttpError(view, request, errorResponse);
+            refresh.setVisibility(View.VISIBLE);
+            if (webViewProgressBar.isShown())
+                webViewProgressBar.setVisibility(View.GONE);
+            //TODO requies api 21
+            //if (errorResponse.getStatusCode())
+            //System.out.println("Error: "+ errorResponse.getStatusCode());
+            Toast.makeText(BrowserActivity.this, "Unexpected HTTP error occurred.Reload page again.", Toast.LENGTH_SHORT).show();
+        }
+
+        //TODO keep this?
+        @Override
+        public void onReceivedSslError(WebView view, SslErrorHandler handler, SslError error) {
+            super.onReceivedSslError(view, handler, error);
+            refresh.setVisibility(View.VISIBLE);
+            if (webViewProgressBar.isShown())
+                webViewProgressBar.setVisibility(View.GONE);
+            Toast.makeText(BrowserActivity.this, "Unexpected SSL error occurred.Reload page again.", Toast.LENGTH_SHORT).show();
+        }
+
+    }
+
+    // To handle "Back" key press event for WebView to go back to previous screen.
+    @Override
+    public boolean onKeyDown(int keyCode, KeyEvent event) {
+        if ((keyCode == KeyEvent.KEYCODE_BACK)) {
+            isWebViewCanGoBack();
+            return true;
+        }
+        return super.onKeyDown(keyCode, event);
+    }
+
+    private void shareUrl(){
+        Intent sendIntent = new Intent();
+        sendIntent.setAction(Intent.ACTION_SEND);
+        sendIntent.putExtra(Intent.EXTRA_TEXT, webView.getUrl());
+        sendIntent.setType("text/plain");
+        startActivity(sendIntent);
+    }
+
+    private void tryMirror(){
+        if ((urlIndex+1) < urls.size()){
```

Comment:
Uhm, it seems to me this function does not handle the case where `urls` is empty.